Mathlib/RingTheory/TwoSidedIdeal/Basic.lean

```diff
@@ -0,0 +1,174 @@
+  coeOrderEmbedding.map_rel_iff.symm
+
+@[ext]
+lemma ext {I J : TwoSidedIdeal R} (h : ∀ x, x ∈ I ↔ x ∈ J) : I = J :=
+  coeOrderEmbedding.injective (Set.ext h)
+
+lemma lt_iff (I J : TwoSidedIdeal R) : I < J ↔ (I : Set R) ⊂ (J : Set R) := by
+  rw [lt_iff_le_and_ne, Set.ssubset_iff_subset_ne, le_iff]
+  simp
+
+lemma zero_mem : 0 ∈ I := I.ringCon.refl 0
+
+lemma add_mem {x y} (hx : x ∈ I) (hy : y ∈ I) : x + y ∈ I := by simpa using I.ringCon.add hx hy
+
+lemma neg_mem {x} (hx : x ∈ I) : -x ∈ I := by simpa using I.ringCon.neg hx
+
+lemma sub_mem {x y} (hx : x ∈ I) (hy : y ∈ I) : x - y ∈ I := by simpa using I.ringCon.sub hx hy
+
+lemma mul_mem_left (x y) (hy : y ∈ I) : x * y ∈ I := by
+  simpa using I.ringCon.mul (I.ringCon.refl x) hy
+
+lemma mul_mem_right (x y) (hx : x ∈ I) : x * y ∈ I := by
+  simpa using I.ringCon.mul hx (I.ringCon.refl y)
+
+lemma nsmul_mem {x} (n : ℕ) (hx : x ∈ I) : n • x ∈ I := by simpa using I.ringCon.nsmul _ hx
+
+lemma zsmul_mem {x} (n : ℤ) (hx : x ∈ I) : n • x ∈ I := by simpa using I.ringCon.zsmul _ hx
+
+/--
+The "set-theoretic-way" of constructing a two-sided ideal by providing:
+- the underlying set `S`;
+- a proof that `0 ∈ S`;
+- a proof that `x + y ∈ S` if `x ∈ S` and `y ∈ S`;
+- a proof that `-x ∈ S` if `x ∈ S`;
+- a proof that `x * y ∈ S` if `y ∈ S`;
+- a proof that `x * y ∈ S` if `x ∈ S`.
+-/
+def mk' (carrier : Set R)
+    (zero_mem : 0 ∈ carrier)
+    (add_mem : ∀ {x y}, x ∈ carrier → y ∈ carrier → x + y ∈ carrier)
+    (neg_mem : ∀ {x}, x ∈ carrier → -x ∈ carrier)
+    (mul_mem_left : ∀ {x y}, y ∈ carrier → x * y ∈ carrier)
+    (mul_mem_right : ∀ {x y}, x ∈ carrier → x * y ∈ carrier) : TwoSidedIdeal R where
+  ringCon :=
+    { r := fun x y => x - y ∈ carrier
+      iseqv :=
+      { refl := fun x => by simpa using zero_mem
+        symm := fun h => by simpa using neg_mem h
+        trans := fun {x y z} h1 h2 => by
+          simpa only [show x - z = (x - y) + (y - z) by abel] using add_mem h1 h2 }
+      mul' := fun {a b c d} (h1 : a - b ∈ carrier) (h2 : c - d ∈ carrier) => show _ ∈ carrier by
+        rw [show a * c - b * d = a * (c - d) + (a - b) * d by rw [mul_sub, sub_mul]; abel]
+        exact add_mem (mul_mem_left h2) (mul_mem_right h1)
+      add' := fun {a b c d} (h1 : a - b ∈ carrier) (h2 : c - d ∈ carrier) => show _ ∈ carrier by
+        rw [show a + c - (b + d) = (a - b) + (c - d) by abel]
+        exact add_mem h1 h2 }
+
+lemma mem_mk' (carrier : Set R)
+    (zero_mem : 0 ∈ carrier)
+    (add_mem : ∀ {x y}, x ∈ carrier → y ∈ carrier → x + y ∈ carrier)
+    (neg_mem : ∀ {x}, x ∈ carrier → -x ∈ carrier)
+    (mul_mem_left : ∀ {x y}, y ∈ carrier → x * y ∈ carrier)
+    (mul_mem_right : ∀ {x y}, x ∈ carrier → x * y ∈ carrier)
+    (x : R) :
+    x ∈ mk' carrier zero_mem add_mem neg_mem mul_mem_left mul_mem_right ↔ x ∈ carrier := by
+  rw [mem_iff]
+  simp [mk']
+
+instance : AddSubgroupClass (TwoSidedIdeal R) R where
+  zero_mem := zero_mem
+  add_mem := @add_mem _ _
+  neg_mem := @neg_mem _ _
```

Comment:
I think you can move this up, just after the `neg_mem` lemma.
Then you can use it in theproofs of `sub_mem` and `nsmul_mem` etc...
Of course the current proofs are also one-liners, so this isn't very important.